# Pytorch Tutorial

Pytorch is a popular deep learning framework and it's easy to get started.

In [2]:
import torch
import torch.nn as nn
import torch.utils.data as data
import torch.nn.functional as F
import torchvision
import torchvision.transforms as transforms
from tqdm import tqdm
import time

BATCH_SIZE = 128
NUM_EPOCHS = 10

First, we read the mnist data, preprocess them and encapsulate them into dataloader form.

In [3]:
# preprocessing
normalize = transforms.Normalize(mean=[.5], std=[.5])
transform = transforms.Compose([transforms.ToTensor(), normalize])

# download and load the data
train_dataset = torchvision.datasets.MNIST(root='./mnist/', train=True, transform=transform, download=True)
test_dataset = torchvision.datasets.MNIST(root='./mnist/', train=False, transform=transform, download=False)

# encapsulate them into dataloader form
train_loader = data.DataLoader(train_dataset, batch_size=BATCH_SIZE, shuffle=True, drop_last=True)
test_loader = data.DataLoader(test_dataset, batch_size=BATCH_SIZE, shuffle=False, drop_last=True)

Then, we define the model, object function and optimizer that we use to classify.

In [4]:
class SimpleNet(nn.Module):
# TODO:define model
    def __init__(self):
        super().__init__()
        self.fun1 = nn.Linear(784,64)
        self.fun2 = nn.Linear(64,64)
        self.fun3 = nn.Linear(64,64)
        self.fun4 = nn.Linear(64,10)
        
    def forward(self, x):
        x = F.relu(self.fun1(x))
        x = F.relu(self.fun2(x))
        x = F.relu(self.fun3(x))
        x = self.fun4(x)
        return F.log_softmax(x, dim=1)

net = SimpleNet()
print (net)
# TODO:define loss function and optimiter
criterion = nn.CrossEntropyLoss()
optimizer = torch.optim.SGD(net.parameters(),lr=0.03)

SimpleNet(
  (fun1): Linear(in_features=784, out_features=64, bias=True)
  (fun2): Linear(in_features=64, out_features=64, bias=True)
  (fun3): Linear(in_features=64, out_features=64, bias=True)
  (fun4): Linear(in_features=64, out_features=10, bias=True)
)


Next, we can start to train and evaluate!

In [5]:
# train and evaluate
for epoch in range(NUM_EPOCHS):
    for data in train_loader:
        
        X, y = data
        net.zero_grad()
        output = net(X.view(-1,28*28))
        #print(output)
        loss = F.nll_loss(output, y)
        loss.backward()
        optimizer.step()
    print(loss)
    correct = 0
    total = 0
    with torch.no_grad():
        for data in train_loader:
            X, y = data
            output = net(X.view(-1,784))
            for idx, i in enumerate(output):
                if torch.argmax(i) == y[idx]:
                    correct += 1
                total += 1
        train_acc = round(correct/total,3)
        print("train_Accuracy: ", train_acc)
    correct = 0
    total = 0
    with torch.no_grad():
        for data in test_loader:
            X, y = data
            output = net(X.view(-1,784))
            for idx, i in enumerate(output):
                if torch.argmax(i) == y[idx]:
                    correct += 1
                total += 1
        test_acc = round(correct/total,3)
        print("test_Accuracy: ", test_acc)
    # evaluate
    # TODO:calculate the accuracy using traning and testing dataset
    
    
    
    


tensor(0.3730, grad_fn=<NllLossBackward>)
train_Accuracy:  0.866
test_Accuracy:  0.867
tensor(0.2521, grad_fn=<NllLossBackward>)
train_Accuracy:  0.906
test_Accuracy:  0.906
tensor(0.3643, grad_fn=<NllLossBackward>)
train_Accuracy:  0.915
test_Accuracy:  0.914
tensor(0.1903, grad_fn=<NllLossBackward>)
train_Accuracy:  0.931
test_Accuracy:  0.931
tensor(0.2007, grad_fn=<NllLossBackward>)
train_Accuracy:  0.938
test_Accuracy:  0.937
tensor(0.2175, grad_fn=<NllLossBackward>)
train_Accuracy:  0.939
test_Accuracy:  0.937
tensor(0.2230, grad_fn=<NllLossBackward>)
train_Accuracy:  0.952
test_Accuracy:  0.948
tensor(0.0766, grad_fn=<NllLossBackward>)
train_Accuracy:  0.959
test_Accuracy:  0.957
tensor(0.2318, grad_fn=<NllLossBackward>)
train_Accuracy:  0.961
test_Accuracy:  0.958
tensor(0.1187, grad_fn=<NllLossBackward>)
train_Accuracy:  0.961
test_Accuracy:  0.958


#### Q5:
Please print the training and testing accuracy.

train_Accuracy:  0.961

test_Accuracy:  0.958